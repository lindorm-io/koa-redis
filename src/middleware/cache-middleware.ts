import { Middleware } from "@lindorm-io/koa";
import { RedisCache } from "@lindorm-io/redis";
import { RedisContext } from "../types";
import { camelCase } from "lodash";

interface CacheMiddlewareOptions {
  key?: string;
  expiresInSeconds?: number;
}

export const cacheMiddleware =
  (Cache: typeof RedisCache, options?: CacheMiddlewareOptions): Middleware<RedisContext> =>
  async (ctx, next): Promise<void> => {
    const start = Date.now();

    /*
     * Ignoring TS here since Cache needs to be abstract
     * to ensure that all input at least attempts to be unique
     */
    // @ts-ignore
    ctx.cache[camelCase(options?.key || Cache.name)] = new Cache({
      client: ctx.client.redis.client(),
      expiresInSeconds: options?.expiresInSeconds,
      logger: ctx.logger,
    });

    ctx.metrics.cache = (ctx.metrics.cache || 0) + (Date.now() - start);

    await next();
  };
