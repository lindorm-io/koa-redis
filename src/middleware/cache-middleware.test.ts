import { cacheMiddleware } from "./cache-middleware";
import { logger } from "../test";

class Test {}

const next = () => Promise.resolve();

describe("cacheMiddleware", () => {
  let ctx: any;

  beforeEach(() => {
    ctx = {
      logger,
      metrics: {},
      redis: { getClient: () => "client" },
      cache: {},
    };
  });

  test("should set cache on context", async () => {
    // @ts-ignore
    await expect(cacheMiddleware(Test)(ctx, next)).resolves.toBe(undefined);
    expect(ctx.cache).toMatchSnapshot();
    expect(ctx.metrics.cache).toStrictEqual(expect.any(Number));
  });

  test("should set cache with specific key", async () => {
    // @ts-ignore
    await expect(cacheMiddleware(Test, { key: "otherKey" })(ctx, next)).resolves.toBe(undefined);
    expect(ctx.cache).toMatchSnapshot();
    expect(ctx.metrics.cache).toStrictEqual(expect.any(Number));
  });
});
