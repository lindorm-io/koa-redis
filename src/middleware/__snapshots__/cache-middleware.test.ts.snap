// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cacheMiddleware should set cache on context 1`] = `
Object {
  "test": Test {},
}
`;

exports[`cacheMiddleware should set cache with specific key 1`] = `
Object {
  "otherKey": Test {},
}
`;
